triggers/service/auth/src/main/scala/com/daml/oauth/middleware/Server.scala

```diff
@@ -0,0 +1,125 @@
+import akka.http.scaladsl.server.Directives._
+import akka.http.scaladsl.unmarshalling.{Unmarshal, Unmarshaller}
+import com.daml.oauth.server.{Request => OAuthRequest, Response => OAuthResponse}
+import com.typesafe.scalalogging.StrictLogging
+import java.util.Base64
+import java.util.UUID
+import scala.collection.concurrent.TrieMap
+import scala.concurrent.{ExecutionContext, Future}
+import scala.language.postfixOps
+import scala.util.Try
+import spray.json._
+
+// This is an implementation of the trigger service authentication middleware
+// for OAuth2 as specified in `/triggers/service/authentication.md`
+object Server extends StrictLogging {
+  import com.daml.oauth.server.JsonProtocol._
+
+  private def toRedirectUri(uri: Uri) =
+    Uri()
+      .withScheme(uri.scheme)
+      .withAuthority(uri.authority)
+      .withPath(Uri.Path./("cb"))
+
+  def start(
+      config: Config)(implicit system: ActorSystem, ec: ExecutionContext): Future[ServerBinding] = {
+    implicit val unmarshal: Unmarshaller[String, Uri] = Unmarshaller.strict(Uri(_))
+    // TODO[AH] Make sure this is bounded in size - or avoid state altogether.
+    val requests = TrieMap[UUID, Uri]()
+    val route = concat(
+      path("auth") {
+        get {
+          complete((StatusCodes.NotImplemented, "The /auth endpoint is not implemented yet"))
+        }
+      },
+      path("login") {
+        get {
+          parameters(('redirect_uri.as[Uri], 'claims))
+            .as[Request.Login](Request.Login) {
+              login =>
+                extractRequest {
+                  request =>
+                    val requestId = UUID.randomUUID
+                    requests += (requestId -> login.redirectUri)
+                    val authorize = OAuthRequest.Authorize(
+                      responseType = "code",
+                      clientId = config.clientId,
+                      redirectUri = toRedirectUri(request.uri),
+                      scope = Some(login.claims),
+                      state = Some(requestId.toString))
+                    redirect(
+                      config.oauthUri
+                        .withPath(Uri.Path./("authorize"))
+                        .withQuery(authorize.toQuery),
+                      StatusCodes.Found)
+                }
+            }
+        }
+      },
+      path("cb") {
+        get {
+          parameters(('code, 'state ?))
+            .as[OAuthResponse.Authorize](OAuthResponse.Authorize) {
+              authorize =>
+                extractRequest {
+                  request =>
+                    val redirectUri = for {
+                      state <- authorize.state
+                      requestId <- Try(UUID.fromString(state)).toOption
+                      redirectUri <- requests.remove(requestId)
+                    } yield redirectUri
+                    redirectUri match {
+                      case None =>
+                        complete(StatusCodes.NotFound)
```

Comment:
Maybe unauthorized would be a more appropriate status code?

Reply:
The idea behind 404 was that the login request under the given `state` was not found. 401 seems problematic, the `/cb` is part of the authentication flow, not something that requires authentication itself, also 401 requires a challenge in the [WWW-Authenticate header field](https://tools.ietf.org/html/rfc7235#section-3.1) which doesn't seem appropriate here.